sql/parser: extract code in different packages

knz opened the pull request:
Supersedes #19922.

tl;dr:
- new package `sem/builtins` with built-in function definitions
- new package `sem/transform` with the expression visitors
- new package `sem/tests` for tests that depend both on `parser` and `builtins`

All split up in miniature commits, for your (re)viewing pleasure. :)

knz committed:
sql/parser: inline isNaN in caller, reduce overhead

knz committed:
sql/parser: new function TimestampDifference, use in builtins

knz committed:
sql/parser: new format flags FmtAlwaysGroupExprs for use by TestParseTree

knz committed:
sql/parser: move the AggregateFunc type definition to its own file

knz committed:
sql/parser: move WindowFunc, WindowFrame and IndexeRow to their own file

knz committed:
sql/parser: move ValueGenerator to its own file

knz committed:
sql/parser: intPow: builtins.go -> eval.go

knz committed:
sql/parser: pickFromTuple: builtins.go -> eval.go

knz committed:
sql/parser: mockNameTypes: type_check.go -> testutils.go

knz committed:
sql/parser: group built-in initialization into a new file all_builtins.go

This moves the functions `init` (from `builtins.go`) and
`collectBuiltins` (from `window_builtins.go`) into a new file.

Also the logic that fills in missing built-in categories, which was
invoked anew every time `Builtin.Category()` was called, is moved to
the one-time `init` function.

knz committed:
sql/parser: move TestCastToCollatedString to its own file

knz committed:
sql/parser: move LIKE and SIMILAR tests to their own file

knz committed:
sql/parser: move TestClusterTimestampConversion to its own file

knz committed:
sql/parser: canBeInArray -> types.IsValidArrayElementType

knz committed:
sql/parser: move EmptyDTable() to generator_builtins.go

knz committed:
sql/parser: move TestHelpFunctions to its own file

knz committed:
sql/parser: move TypeCheck and TypeCheckError to their own file

knz committed:
sql/sem/tests: new test folder for tests that need multiple packages

This folder is intended for use by tests from the `sql/parser` package
that have dependencies on stuff from both `parser` and other packages
that will be split off `parser` but depend on it.

A separate test directory is needed to avoid errors from the Go
compiler about dependency cycles.

The directory is initially populated with stub aliases from the
`parser` package so that git more readily recognizes file moves.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19941)
<!-- Reviewable:end -->

knz committed:
pkg/sql: move format_test.go to sem/tests

knz committed:
sql/sem/tests: more temporary stubs

knz committed:
pkg/sql: move function_name_test.go to sem/tests

knz committed:
sql/sem/tests: more temporary stubs

knz committed:
pkg/sql: move timestamp_test.go to sem/tests

knz committed:
sql/sem/tests: more temporary stubs

knz committed:
pkg/sql: move parse_test.go to sem/tests

knz committed:
sql/sem/tests: more temporary stubs

knz committed:
pkg/sql: move normalize_test.go to sem/tests

knz committed:
sql/parser: move TestEvalComparisonExprCaching to its own file

knz committed:
sql/sem/tests: more temporary stubs

knz committed:
pkg/sql: move eval_test.go to sem/tests

knz committed:
sql/sem/tests: more temporary stubs

knz committed:
pkg/sql: move full_type_check_test.go to sem/tests (rename to type_check_test.go)

knz committed:
sql/sem/tests: remove the alias stubs

knz committed:
sql/parser: move the expression property visitors to a new package sem/transforms

knz committed:
sql/parser: move FunctionClass and Builtin to their own file

knz committed:
sql/parser: move the built-in function definitions to sem/builtins

This in turn requires a few new API exports:

- the fields of `Builtin`. This in turn removes the
  need for its public accessors.
- `AddWithOverflow`
- `AppendToMaybeNullArray`
- `ConcatArrays`
- `ErrDivByZero`
- `ErrZeroModulus`
- `EvalContext.StmtTimestamp`
- `EvalContext.TxnTimestamp`
- `FixedReturnType`, `UnknownReturnType`, `IdentityReturnType`
- `FunDefs`
- `IntPow`
- `NewFunctionDefinition`
- `PickFromTuple`
- `PrependToMaybeNullArray`
- `RegexpCacheKey` and its `Pattern` interface method.
- `ReturnTyper`
- `SecondsInDay`
- `TypeList`

knz committed:
sql/sem/builtins: new package for built-in function definitions

This package is intended to contain the built-in function definitions,
away from package `sql/parser`. The Go types to hold built-in
functions (`parser.Builtin` etc) will remain in `parser` for the time
being.

The directory is initially populated with stub aliases from the
`parser` package so that git more readily recognizes file moves.

knz committed:
sql/sem/builtins: remove the alias stubs

knz committed:
sql/parser: move non-parser code to a new package sem/tree

However the tests that still depend on the Parse() functions
are left in the `parser` package.

knz committed:
sql/parser: break the dependency eval.go -> parse.go

knz committed:
sql/parser: do not use the StrVal constructor directly in sql.y

knz committed:
sql/parser: export Tuple.Row

knz committed:
sql/parser: export the TableDef.SetName method

knz committed:
sql/parser: export (*TransactionModes).Merge()

knz committed:
sql/parser: export the ColumnKeyword field of the AlterTable nodes

knz committed:
sql/parser: export the SyntaxMode field of Cast and AnnotateType expr nodes

knz committed:
sql/parser: export SelectClause.TableSelect

knz committed:
sql/parser: only use the exported API in constant_test.go

knz committed:
sql/parser: move stripMemoizedFuncs to type_check.go

knz committed:
sql/parser: avoid adding a new method to TableName in a test

knz committed:
sql/parser: export Datum's min and max methods

knz committed:
sem/tree: remove the stub aliases

knz committed:
sql/parser: remove the lowercase (local) stubs

knz committed:
sql/parser: remove the alias stubs

knz committed:
sql/sem/tests, sql/parser: move the tests to their rightful location

jordanlewis commented on pkg/sql/parser/expr_test.go:
```diff
@@ -151,7 +151,7 @@ func TestExprString(t *testing.T) {
 		`(1 >= 2) IS OF (BOOL)`,
 		`(1 >= 2) = (2 IS OF (BOOL))`,
-		`count(1) FILTER (WHERE true)`,
+		// `count(1) FILTER (WHERE true)`,
```
Second to last commit does this for some reason, probably a mistake

jordanlewis commented:
LGTM! :tada: (Reviewable and GitHub's review view are both unusably slow on this size of PR so I can't leave top level comments there.)

Overall I'm *very* happy with this new structure. Thank you for taking on this large burden! I admit to not having read every single line of code change, but I feel confident for the record that this is 98% restructuring.

I saw that you added new exported fields to `Builtin`, but left the accessor methods unexported. This seems backwards - if there's a good reason please chime in but if not I'd prefer to see the opposite done.

Also if we end up doing much more iteration on this PR please split this PR into one before the builtins change and one containing just the builtins change.

I would suggest adding a readme to the new `sem/tests` directory explaining what you did in the commit message.

I think our test structuring could still use some work - it's always been hard to figure out where a test lives or where a new one should live, and it's not clear that it is getting easier - but that's not really the responsibility of this PR to fix. Just putting it for the record.

knz commented:
TFYR!

The Builtin methods you have seen are only visible to TypeCheck, and thus need not be exported.

knz committed:
sql/sem/tests: re-enable the built-in test in TestExprString

knz committed:
sql/sem/tests: add README.md to explain purpose
